### PR TITLE
feat: add new from params to prover

### PR DESCRIPTION
**Pull request by dongchangYoo**
The existing `TachyonSHPlonkProver` and `TachyonGWCProver` were initiated through `unsafe_setup` after receiving a secret (tau) value as the input. With this PR, the `Prover`s can now be initiated with pre-generated params as the input.

This PR should be merged after https://github.com/kroma-network/tachyon/pull/353

### CODE REVIEW

File: tachyon/c/zk/plonk/halo2/buffer_reader.h [hunk 1]
```diff
@@ -405,6 +407,43 @@ class BufferReader<tachyon::zk::plonk::PermutationProvingKey<Poly, Evals>> {
...
+          return point;
+        });
+    std::vector<G1Point> g1_powers_of_tau_lagrange =
+        tachyon::base::CreateVector(n, [&buffer]() {
+          G1Point point;
+          c::zk::plonk::ReadBuffer(buffer, point);
+          return point;
+        });
+
+    // NOTE(dongchangYoo): read |g2| but not use it.
```
Comment (ashjeong) [inline tachyon/c/zk/plonk/halo2/buffer_reader.h:436]:
```suggestion
    // NOTE(dongchangYoo): read |g2| but do not use it.

**Review by ashjeong — CHANGES_REQUESTED** [review]
https://github.com/kroma-network/tachyon/pull/355/commits/db7a0908d47bbc7d02a18304aced4cad28dfccbd

refac(c/zk): move `buffer_reader` to c/zk/plonk/halo2/ 
->
refac(c/zk): move `buffer_reader` to `c/zk/plonk/halo2/` (change directory to code snippet)

https://github.com/kroma-network/tachyon/pull/355/commits/0daa0ef4ee5576b69d7f9ab79589713323f1056a
Don’t start a sentence with “But” or “And”

Unit tests for each `tachyon_halo2_bn254_gwc_prover_create_from_params` and `tachyon_halo2_bn254_shplonk_prover_create_from_params` was added.

And `ExpectedHalo2Params` in `bn254_halo2_params_data.h` can be derived from the `Halo2`.
->
Added unit tests for `tachyon_halo2_bn254_gwc_prover_create_from_params`
and `tachyon_halo2_bn254_shplonk_prover_create_from_params`.

`ExpectedHalo2Params` in `bn254_halo2_params_data.h` can now be derived from `Halo2`.

https://github.com/kroma-network/tachyon/pull/355/commits/77addd713f6df88a8f01d2cb7cb9b2d7ebe69bd7
From now, `TachyonSHPlonkProver` and `TachyonGWCProver` can be initiated with `params`.
->
`TachyonSHPlonkProver` and `TachyonGWCProver` can now be initiated with `params`.

File: tachyon/c/zk/plonk/halo2/test/bn254_halo2_params_data.h [hunk 2]
```diff
@@ -0,0 +1,304 @@
+#ifndef TACHYON_C_ZK_PLONK_HALO2_TEST_BN254_HALO2_PARAMS_DATA_H_
+#define TACHYON_C_ZK_PLONK_HALO2_TEST_BN254_HALO2_PARAMS_DATA_H_
+
+#include <stddef.h>
```
Comment (TomTaehoonKim) [inline tachyon/c/zk/plonk/halo2/test/bn254_halo2_params_data.h:4]: I'm not sure if this header is being used.

**Comment by TomTaehoonKim** [comment]
@dongchangYoo How about fixing commit message for 0daa0ef?

`feat(c/zk): add method to create c-prover from params`

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/c/zk/plonk/halo2/test/bn254_halo2_params_data.h (hunk 2 above)
Comment (dongchangYoo) [reply tachyon/c/zk/plonk/halo2/test/bn254_halo2_params_data.h:4]: The header is for `size_t`.

**Comment by chokobole** [comment]
Please add the scope `(c/zk)` to c7c89cfed614303ba1ddd019ba384fb71e49b44c

File: tachyon/c/zk/plonk/halo2/test/bn254_halo2_params_data.h (hunk 2 above)
Comment (dongchangYoo) [reply tachyon/c/zk/plonk/halo2/test/bn254_halo2_params_data.h:4]: I've confirmed that `size_t` is not used in this file!

**Review by ashjeong — APPROVED** [review]
LGTM

File: tachyon/c/zk/plonk/halo2/buffer_reader.h [hunk 3]
```diff
@@ -405,6 +407,43 @@ class BufferReader<tachyon::zk::plonk::PermutationProvingKey<Poly, Evals>> {
...
+          c::zk::plonk::ReadBuffer(buffer, point);
+          return point;
+        });
+
+    // NOTE(dongchangYoo): read |g2| but do not use it.
+    G2Point g2;
+    c::zk::plonk::ReadBuffer(buffer, g2);
+
+    G2Point s_g2;
+    c::zk::plonk::ReadBuffer(buffer, s_g2);
```
Comment (Insun35) [inline tachyon/c/zk/plonk/halo2/buffer_reader.h:441]: I guess you don't need `c::zk::plonk` cuz its in the same namespace.
```suggestion
	  ReadBuffer(buffer, point);
          return point;
        });
    std::vector<G1Point> g1_powers_of_tau_lagrange =
        tachyon::base::CreateVector(n, [&buffer]() {
          G1Point point;
          ReadBuffer(buffer, point);
          return point;
        });

    // NOTE(dongchangYoo): read |g2| but do not use it.
    G2Point g2;
    ReadBuffer(buffer, g2);

    G2Point s_g2;
    ReadBuffer(buffer, s_g2);
```

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by chokobole — APPROVED** [review]
LGTM

**Pull request merged**